{% extends '../layout.html' %} {% load static %} {% block content %}
<section id="exceptions_assertions">
	<h4>Exceptions and Assertions in Python</h4>
	<p>
		<span class="italicsText">Exceptions</span> and
		<span class="italicsText">assertions</span> in Python provide additional
		means of defensive programming.
	</p>
	<p>
		<span class="topic">Exceptions.</span> Exceptions in Python directly address
		the question: What happens when the procedure is executed and meets an
		unexpected condition? Hitting an unexpected condition is an
		<span class="term">exception</span>. When an exception occurs, we typically
		get back error messages (e.g., <span class="monoText">IndexError</span>,
		<span class="monoText">TypeError</span>,
		<span class="monoText">NameError</span>, etc.).
	</p>
	<p>
		Other common types exceptions: <span class="monoText">SyntaxError</span> -
		Python cannot parse the program. <span class="monoText">NameError</span> - a
		local or global name cannot be seen.
		<span class="monoText">AttributeError</span> - an attribute reference has
		failed. <span class="monoText">TypeError</span> - the operand does not have
		the correct type. <span class="monoText">ValueError</span> - the operand
		type is valid, but the value is illegal.
		<span class="monoText">IOError</span> - There is a problem with the IO
		system (e.g., nonexistent file or file not found).
	</p>
	<p>
		<span class="topic">What do we do with exceptions?</span> There are three
		options for handling exceptions: (1) Fail silently; (2) Return an
		&#8220;error&#8221; value; (3) stop execution and signal the condition.
	</p>
	<p>
		The first option, failing silently, is where we substitute default values or
		simply continue. This is the worst possible approach. It is essentially
		sweeping the problem under the rug and pretending it does not exist. Worse,
		it borders on deceit. The user will get no warning that something has gone
		wrong.
	</p>
	<p>
		The second option, return an error value, is an incomplete solution. Even if
		we return an error value, we still have the problem. This can also
		needlessly complicate our code. What value should we choose? How do we check
		for the special value?
	</p>
	<p>
		The third option is the one we should always go with &mdash; as soon as the
		code hits an exception, stop executing and
		<span class="italicsText">raise an exception</span>. This is done in Python
		with the following:
	</p>
	<pre class="language-python"><code>
		raise Exception("string describing error")
	</code></pre>
	<p>
		<span class="topic">Handlers.</span> Python provides symbols for handling
		exceptions. Consider the following:
	</p>
	<pre class="language-python"><code>
		try:
			a = int(input("Divisor: "))
			b = int(input("Dividend: "))
			print(a/b)
			print("Okay")
		except:
			print("Bug in user input")
		print("Outside")
	</code></pre>
	<p>
		The code above contains two new pieces of syntax:
		<span class="monoText">try</span> and <span class="monoText">except</span>.
		The code effectively provides, <span class="monoText">try</span> these
		statements below; if at any point an exception is raised, STOP, jump to the
		statement below <span class="monoText">except</span>.
	</p>
	<p>
		The <span class="monoText">try</span> and
		<span class="monoText">except</span> structure above essentially allows us
		to customize exception handling.
	</p>
	<p>
		<span class="topic">Handling Specific Exceptions.</span> Handlers can be
		extended to specific errors. For example:
	</p>
	<pre class="language-python line-number"><code>
		try:
			a = int(input("First term: "))
			b = int(input("Second term: "))
			print("a/b = ", a/b)
			print("a+b = ", a+b)
		except ValueError:
			print("Could not cast to a number.")
		except ZeroDivisionError:
			print("Can't divide by zero.")
		except:
			print("Something went wrong.")	
	</code></pre>
	<p>
		The code above provides that if the exceptions above occur, stop execution
		and go to the relevant exception. Instead of print out the usual error
		message (i.e., <span class="monoText">ValueError</span>) do this instead.
	</p>
</section>
{% endblock %}
